### ISSUE TITLE
fill-pattern does not display when changing patterns using "step", "zoom" 

### ISSUE THREAD

**Issue by sequel63** [maplibre/maplibre-gl-js#3106]
Using different fill-patterns and switching them does not display any pattern at step zoom level.

"fill-pattern": [
            "step", ["zoom"],
            "zoo_11", 4,
            "volcano_11"
        ]
    }

**maplibre-gl-js version**: 2.4.0

**browser**: Chrome

### Steps to Trigger Behavior

 1. https://jsbin.com/cuxetuxexu/edit?html,output
 2. Zoom gradually to sea area.
 3. At Zoom level 4 sea fill (zoo) disappears.
 4. At Zoom level 5 new symbols (volcanoes) appear correctly.
 5. Subsequent zooming in/out does not replicate the issue, only on initial load and zoom.

### Link to Demonstration
https://jsbin.com/cuxetuxexu/edit?html,output

### Expected Behavior
Zoom level 4 should have zoo fill pattern.
### Actual Behavior
Zoom level 4 has no fill pattern.

**Comment by HarelM** [comment]
Interesting... Looks likes some rounding issue probably...
Feel free to investigate and open a PR to fix this.

**Comment by sequel63** [comment]
Don't have skill required to investigate unfortunately. Willing to put a moderate bounty on this to help expedite...

**Comment by 1ec5** [comment]
@msbarry also noticed this issue in mapbox/mapbox-gl-js#9724.

**Comment by CraigglesO** [comment]
@sequel63 - are you still interested in expediting this problem? I might have time to solve this.

**Comment by sequel63** [comment]
@CraigglesO yes that would be great. Not familiar with how I can put a bounty on this for you, but if you can assist it would be appreciated!

**Comment by HarelM** [comment]
I would recommend connecting directly and negotiate the terms :-)

**Comment by CraigglesO** [comment]
Cool. Well, I solved the bug. I'm pretty happy with the way I resolved this as well.

@sequel63 -
my discord: craiggleso
slack via OpenStreetMap: Craig OConnor
If you would like to connect and share what you're willing to pay that would be great, we can resolve this ASAP.

@HarelM - I ONLY address this bug and I do it in a way that fixes the bug but the whole area of code is actually broken and problematic. I hope it's ok that although I fix this bug and specific symptoms, that the real fundamental issue is unaddressed? I can explain more when I create a PR.
